### PR TITLE
Issue #26969:prevent avg costed neg balance by period

### PR DESCRIPTION
**Pull request by garyhgohoos**

### CODE REVIEW

**Comment by rkannadevara** [comment]
Performed UAT: Issue is fixed.

File: foundation-database/public/functions/postintoinvbalance.sql [hunk 1]
```diff
@@ -147,6 +146,16 @@ BEGIN
...
           true );
     END IF;
+
+--  Validate negative balance for average costed item
+    IF EXISTS(SELECT invbal_id
+              FROM invbal JOIN itemsite ON (itemsite_id=invbal_itemsite_id)
+              WHERE (invbal_id=_invbalid)
+                AND (invbal_qoh_ending < 0.0)
+                AND (itemsite_costmethod='A')) THEN
+      RAISE EXCEPTION 'Average costed Item with negative balance for invhist_id %, transaction date %.  Transaction can not be posted.', _r.invhist_id, formatDate(_r.invhist_transdate);
```
Comment (gilmoskowitz) [inline foundation-database/public/functions/postintoinvbalance.sql]: Please add the `[xtuple: postIntoInvBalance, ...]` suffix to the message.

File: foundation-database/public/functions/postinvhist.sql [hunk 2]
```diff
@@ -1,36 +1,30 @@
...
-    IF ( SELECT metric_value
-        FROM metric
-        WHERE ((metric_name = 'EnableAsOfQOH')
-        AND (metric_value = 't'))) THEN
-      IF (NOT postIntoInvBalance(pInvhistId)) THEN
-        RAISE EXCEPTION 'Post into Inventory Balance for invhist_id=% was unsuccessful',pInvhistId;
-      END IF;
+  IF ( fetchMetricBool('EnableAsOfQOH') ) THEN
+    IF (NOT postIntoInvBalance(pInvhistId)) THEN
+      RAISE EXCEPTION 'Post into Inventory Balance for invhist_id=% was unsuccessful',pInvhistId;
```
Comment (gilmoskowitz) [inline foundation-database/public/functions/postinvhist.sql]: Please add the `[xtuple: postInvHist, ...]` suffix to the message.

**Comment by Panduk** [comment]
Performed UAT: Observed the following message

**Comment by gilmoskowitz** [comment]
This form of the error message is OK for now (`A Stored Procedure failed to run properly`). We will resolve this as part of the error handling refactoring project currently underway.

**Pull request merged**